crates/vm/levm/src/db/gen_db.rs

```diff
@@ -61,27 +61,39 @@ impl GeneralizedDatabase {
 impl<'a> VM<'a> {
     // ================== Account related functions =====================
 
-    pub fn get_account_mut(&mut self, address: Address) -> Result<&mut Account, VMError> {
-        let backup_account = match cache::get_account(&self.db.cache, &address) {
-            Some(acc) => acc.clone(),
-            None => {
-                let acc = self.db.store.get_account(address)?;
-                cache::insert_account(&mut self.db.cache, address, acc.clone());
-                acc
-            }
-        };
+    /*
+        Each callframe has a CallFrameBackup, which contains:
 
-        if let Ok(frame) = self.current_call_frame_mut() {
-            frame
-                .cache_backup
-                .entry(address)
-                .or_insert_with(|| Some(backup_account));
-        }
+        - A list with account infos of every account that was modified so far (balance, nonce, bytecode/code hash)
+        - A list with a tuple (address, storage) that contains, for every account whose storage was accessed, a hashmap
+        of the storage slots that were modified, with their original value.
 
-        let account = cache::get_account_mut(&mut self.db.cache, &address)
-            .ok_or(VMError::Internal(InternalError::AccountNotFound))?;
+        On every call frame, at the end one of two things can happen:
 
-        Ok(account)
+        - The transaction succeeds. In this case:
+            - The CallFrameBackup of the current callframe has to be merged with the backup of its parent, in the following way:
+            For every account that's present in the parent backup, do nothing (i.e. keep the one that's already there).
+            For every account that's NOT present in the parent backup but is on the child backup, add the child backup to it.
+            Do the same for every individual storage slot.
+        - The transaction reverts. In this case:
+            - Insert into the cache the value of every account on the CallFrameBackup.
+            - Insert into the cache the value of every storage slot in every account on the CallFrameBackup.
+
+    */
+    pub fn get_account_mut(&mut self, address: Address) -> Result<&mut Account, VMError> {
+        if cache::is_account_cached(&self.db.cache, &address) {
+            self.backup_account_info(address)?;
+            cache::get_account_mut(&mut self.db.cache, &address).ok_or(VMError::Internal(
+                crate::errors::InternalError::AccountNotFound,
+            ))
+        } else {
+            let acc = self.db.store.get_account(address)?;
+            cache::insert_account(&mut self.db.cache, address, acc.clone());
```

Comment:
```suggestion
            cache::insert_account(&mut self.db.cache, address, acc);
```
If I'm not wrong it seems this `.clone()` can be removed

Reply:
That's correct, good catch